demmt/object_gf100_compute.c

```diff
@@ -0,0 +1,157 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "buffer.h"
+#include "config.h"
+#include "demmt.h"
+#include "log.h"
+#include "nvrm.h"
+#include "object.h"
+
+struct gf100_compute_data
+{
+	struct addr_n_buf code;
+	struct addr_n_buf cb;
+	int data_offset;
+
+	struct mthd2addr *addresses;
+};
+
+static void destroy_gf100_compute_data(struct gpu_object *obj)
+{
+	struct gf100_compute_data *d = obj->class_data;
+	free(d->addresses);
+	free(d);
+}
+
+void decode_gf100_compute_init(struct gpu_object *obj)
+{
+	struct gf100_compute_data *d = obj->class_data = calloc(1, sizeof(struct gf100_compute_data));
+	obj->class_data_destroy = destroy_gf100_compute_data;
+
+#define SZ 3
+	struct mthd2addr *tmp = d->addresses = calloc(SZ, sizeof(*d->addresses));
+	m2a_set1(tmp++, 0x1608, 0x160c, &d->code);
+	m2a_set1(tmp++, 0x2384, 0x2388, &d->cb);
+	m2a_set1(tmp++, 0, 0, NULL);
+	assert(tmp - d->addresses == SZ);
+#undef SZ
+}
+
+void decode_gf100_compute_terse(struct gpu_object *obj, struct pushbuf_decode_state *pstate)
+{
+	struct gf100_compute_data *objdata = obj->class_data;
+
+	if (check_addresses_terse(pstate, objdata->addresses))
+	{ }
+}
+
+void decode_gf100_compute_verbose(struct gpu_object *obj, struct pushbuf_decode_state *pstate)
+{
+	int mthd = pstate->mthd;
+	uint32_t data = pstate->mthd_data;
+	struct gf100_compute_data *objdata = obj->class_data;
+
+	if (check_addresses_verbose(pstate, objdata->addresses))
+	{ }
+	else if (mthd == 0x1698) // FLUSH
+	{
+		int flags_ok = (data & 0x1) == 0x1 ? 1 : 0;
+		mmt_debug("flush: 0x%08x code: %d\n", data, flags_ok);
+
+		if (!flags_ok || objdata->cb.gpu_mapping == NULL)
+		{
+			objdata->cb.address = 0;
+			objdata->cb.gpu_mapping = NULL;
+		}
+	}
+	else if (mthd == 0x238c) // CB_POS
+	{
+		mmt_debug("cb_pos: 0x%08x\n", data);
+		if (objdata->cb.gpu_mapping)
+		{
+			objdata->data_offset = data;
+		}
+	}
+	else if (mthd == 0x2390) // CB_DATA[0]
+	{
+		mmt_debug("cb_data: 0x%08x\n", data);
+		if (objdata->cb.gpu_mapping)
+		{
+			gpu_mapping_register_write(objdata->cb.gpu_mapping,
+						   objdata->cb.address + objdata->data_offset,
+						   4, &data);
+			objdata->data_offset += 4;
+		}
+	}
+	else if (mthd == 0x03b4) // CP_START_ID
+	{
+		struct varinfo *var;
+
+		if (!isa_gf100)
+			isa_gf100 = ed_getisa("gf100");
+
+		var = varinfo_new(isa_gf100->vardata);
+		varinfo_set_variant(var, "gf100");
+
+		struct region *reg;
+		uint8_t *code = NULL;
+		uint64_t code_addr = objdata->code.address;
+		uint32_t start_id = data;
+		struct gpu_mapping *m = objdata->code.gpu_mapping;
+		if (!m) {
```

Comment:
Do you need to? Can you just use objdata->code.gpu_mapping directly and not worry about the if (!m) case?